ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,409 @@
+            "y_ignore": y_ignore,
+            "y_pred_ignore": y_pred_ignore,
+        }
+
+    def _accumulate(self) -> None:
+        num_iou_thr = len(self.iou_thresholds)
+        num_rec_thr = len(self.rec_thresholds)
+        num_categories = len(self.category_ids)
+        num_area = len(self.object_area_ranges)
+
+        max_det = self.num_detection_max
+        precision = -torch.ones((num_iou_thr, num_rec_thr, num_categories, num_area), device=self._device)
+
+        # retrieve eval_imgs at each category, area range, and max number of detections
+        for c, category_id in enumerate(self.category_ids):
+            for a, area_rng in enumerate(self.object_area_ranges):
+                # retrieve appropriate eval_imgs from stored results
+                eval_imgs = self.eval_imgs[category_id, area_rng]
+                eval_imgs = [img for img in eval_imgs if img is not None]
+                if len(eval_imgs) == 0:
+                    continue
+                # Get prediction scores to greedily match
+                pred_scores = torch.cat([img["y_pred_scores"][0:max_det] for img in eval_imgs], dim=-1)
+                # Sort prediction scores
+                inds = torch.argsort(-pred_scores)
+                # Retrieve and Sort prediction matches,
+                # ignore flags for ground truth and predictions based on prediction scores
+                predm = torch.cat([img["y_pred_matches"][:, 0:max_det] for img in eval_imgs], dim=-1)[:, inds]
+                pred_ignore = torch.cat([img["y_pred_ignore"][:, 0:max_det] for img in eval_imgs], dim=-1)[:, inds]
+                y_ignore = torch.cat([img["y_ignore"] for img in eval_imgs])
+                non_ignored = torch.count_nonzero(y_ignore == 0)
+                if non_ignored == 0:
+                    continue
+                # Calculate true positive and false positive based on prediction matches and ignore flags
+                tps = torch.logical_and(predm, torch.logical_not(pred_ignore))
+                fps = torch.logical_and(torch.logical_not(predm), torch.logical_not(pred_ignore))
+                # Calculate precision and recall iteratively
+                tp_sum = torch.cumsum(tps, dim=1).to(device=self._device, dtype=torch.float64)
+                fp_sum = torch.cumsum(fps, dim=1).to(device=self._device, dtype=torch.float64)
+
+                for t, (tp, fp) in enumerate(zip(tp_sum, fp_sum)):
+                    nd = len(tp)
+                    rc = tp / non_ignored
+                    pr = tp / (fp + tp + torch.finfo(torch.float64).eps)
+                    q: Any = torch.zeros((num_rec_thr,))
+
+                    pr = pr.tolist()
+                    q = q.tolist()
+
+                    for i in range(nd - 1, 0, -1):
+                        if pr[i] > pr[i - 1]:
+                            pr[i - 1] = pr[i]
+
+                    inds = torch.searchsorted(rc, self.rec_thresholds, right=False)
+                    try:
```

Comment:
The exception seems weird. Could you explain ?

Reply:
If the recording threshold(self.rec_thresholds) doesn't exist in the recall(rc) tensor, it will return len(rc) + 1 for that value. So we use try and except to skip past that value.